File: gestalt-core/src/main/java/org/config/gestalt/decoder/ByteDecoder.java
package org.config.gestalt.decoder;

import org.config.gestalt.entity.ValidationError;
import org.config.gestalt.node.ConfigNode;
import org.config.gestalt.reflect.TypeCapture;
import org.config.gestalt.utils.ValidateOf;

import java.nio.charset.Charset;

public class ByteDecoder extends LeafDecoder<Byte> {

    @Override
    public Priority priority() {
        return Priority.MEDIUM;
    }

    @Override
    public String name() {
        return "Byte";
    }

    @Override
    public boolean matches(TypeCapture<?> klass) {
        return klass.isAssignableFrom(Byte.class) || klass.isAssignableFrom(byte.class);
    }

    @Override
    protected ValidateOf<Byte> leafDecode(String path, ConfigNode node) {
        ValidateOf<Byte> results;

        String value = node.getValue().orElse("");
        if (value.length() == 1) {
            results = ValidateOf.valid(value.getBytes(Charset.defaultCharset())[0]);
        } else {
            results = ValidateOf.inValid(new ValidationError.DecodingByteTooLong(path, node));
        }

        return results;
    }
}

File: gestalt-kotlin/src/main/kotlin/org/config/gestalt/kotlin/decoder/ByteDecoder.kt
package org.config.gestalt.kotlin.decoder

import org.config.gestalt.decoder.LeafDecoder
import org.config.gestalt.decoder.Priority
import org.config.gestalt.entity.ValidationError
import org.config.gestalt.kotlin.reflect.KTypeCapture
import org.config.gestalt.node.ConfigNode
import org.config.gestalt.reflect.TypeCapture
import org.config.gestalt.utils.ValidateOf
import java.nio.charset.Charset

class ByteDecoder : LeafDecoder<Byte>() {
    override fun name(): String {
        return "Byte"
    }

    override fun priority(): Priority {
        return Priority.HIGH
    }

    override fun matches(klass: TypeCapture<*>): Boolean {
        return if (klass is KTypeCapture<*>) {
            klass.isAssignableFrom(Byte::class)
        } else {
            false
        }
    }

    override fun leafDecode(path: String?, node: ConfigNode): ValidateOf<Byte> {
        val results: ValidateOf<Byte>
        val value = node.value.orElse("")
        results = if (value.length == 1) {
            ValidateOf.valid(value.toByteArray(Charset.defaultCharset())[0])
        } else {
            ValidateOf.inValid(
                ValidationError.DecodingByteTooLong(
                    path,
                    node
                )
            )
        }
        return results
    }
}
